## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[["Attrition", "Department"]]
y_df.head()

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [4]:
# Create a list of at least 10 column names to use as X data
ten_columns = ["Education", "Age", "DistanceFromHome", "JobSatisfaction", "OverTime",
                  "StockOptionLevel", "WorkLifeBalance", "YearsAtCompany", "YearsSinceLastPromotion", "NumCompaniesWorked"]


# Create X_df using your selected columns
X_df = attrition_df[ten_columns]

# Show the data types for X_df
X_df.dtypes


Education                   int64
Age                         int64
DistanceFromHome            int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
dtype: object

In [5]:
X_df.head()

Education  Age  DistanceFromHome  JobSatisfaction OverTime  \
0          2   41                 1                4      Yes   
1          1   49                 8                2       No   
2          2   37                 2                3      Yes   
3          4   33                 3                3      Yes   
4          1   27                 2                2       No   

   StockOptionLevel  WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0                 0                1               6                        0   
1                 1                3              10                        1   
2                 0                3               0                        0   
3                 0                3               8                        3   
4                 1                3               2                        2   

   NumCompaniesWorked  
0                   8  
1                   1  
2                   6  
3                   1  
4                   9

In [6]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=42)

In [7]:
# Function for converting OverTime column to int
def convert_overtime(data):
    data["OverTime"] = data["OverTime"].map({"Yes": 1, "No": 0}).astype("int")
    return data

In [8]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_train = convert_overtime(X_train)
X_test = convert_overtime(X_test)

X_train["OverTime"].value_counts()

OverTime
0    780
1    322
Name: count, dtype: int64

In [9]:
X_train.dtypes

Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object

In [10]:
X_train.head()

Education  Age  DistanceFromHome  JobSatisfaction  OverTime  \
1343          3   29                 7                1         0   
1121          4   36                 1                3         0   
1048          3   34                 3                1         0   
1393          3   27                 9                4         0   
527           3   32                10                4         0   

      StockOptionLevel  WorkLifeBalance  YearsAtCompany  \
1343                 0                3               3   
1121                 0                3               1   
1048                 0                3              13   
1393                 0                3               7   
527                  0                2              10   

      YearsSinceLastPromotion  NumCompaniesWorked  
1343                        1                   3  
1121                        0                   6  
1048                        3                   3  
1393                        0                   1  
527                         0                   1

In [11]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [12]:
# Create a OneHotEncoder for the Department column
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder()

# Fit the encoder to the training data
ohe.fit(y_train[["Department"]])

# Create two new variables by applying the encoder
# to the training and testing data
department_train_encoded = ohe.transform(y_train[["Department"]]).toarray()
department_test_encoded = ohe.transform(y_test[["Department"]]).toarray()

department_test_encoded



array([[0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.],
       ...,
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.]])

In [13]:
# Create a OneHotEncoder for the Attrition column
ohe = OneHotEncoder()

# Fit the encoder to the training data
ohe.fit(y_train[["Attrition"]])

# Create two new variables by applying the encoder
# to the training and testing data
attrition_train_encoded = ohe.transform(y_train[["Attrition"]]).toarray()
attrition_test_encoded = ohe.transform(y_test[["Attrition"]]).toarray()

attrition_test_encoded[:10]


array([[1., 0.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.]])

## Create, Compile, and Train the Model

In [14]:
import tensorflow as tf
# Find the number of columns in the X training data
input_nodes = len(X_train.columns)

# Create the input layer
input_layer = layers.Input(shape=(input_nodes,), name="input")

# Create at least two shared layers
shared1 = layers.Dense(64, activation="relu", name="shared1")(input_layer)
shared2 = layers.Dense(128, activation="relu", name="shared2")(shared1)

In [15]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(32, activation="relu", name="department_hidden")(shared2)

# Create the output layer
department_output = layers.Dense(3, activation="softmax", name="department_output")(department_hidden)


In [16]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(32, activation="relu", name="attrition_hidden")(shared2)

# Create the output layer
attrition_output = layers.Dense(2, activation="sigmoid", name="attrition_output")(attrition_hidden)


In [17]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output], name="model")

# Compile the model
model.compile(optimizer="adam",
              loss={"department_output": "categorical_crossentropy", "attrition_output": "binary_crossentropy"},
              metrics={"department_output": "accuracy", "attrition_output": "accuracy"})

# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input (InputLayer)          [(None, 10)]                 0         []                            
                                                                                                  
 shared1 (Dense)             (None, 64)                   704       ['input[0][0]']               
                                                                                                  
 shared2 (Dense)             (None, 128)                  8320      ['shared1[0][0]']             
                                                                                                  
 department_hidden (Dense)   (None, 32)                   4128      ['shared2[0][0]']             
                                                                                              

In [18]:
# Train the model

model.fit(
    X_train_scaled,
    {"department_output": department_train_encoded, "attrition_output": attrition_train_encoded},
    epochs=100,
    batch_size=35,
    validation_split=0.2
)

Epoch 1/100
26/26 [==============================] - 3s 31ms/step - loss: 1.4438 - department_output_loss: 0.8619 - attrition_output_loss: 0.5820 - department_output_accuracy: 0.5925 - attrition_output_accuracy: 0.7435 - val_loss: 1.4104 - val_department_output_loss: 0.8857 - val_attrition_output_loss: 0.5246 - val_department_output_accuracy: 0.6063 - val_attrition_output_accuracy: 0.7873
Epoch 2/100
26/26 [==============================] - 0s 12ms/step - loss: 1.2107 - department_output_loss: 0.7703 - attrition_output_loss: 0.4405 - department_output_accuracy: 0.6652 - attrition_output_accuracy: 0.8388 - val_loss: 1.3523 - val_department_output_loss: 0.8609 - val_attrition_output_loss: 0.4914 - val_department_output_accuracy: 0.6063 - val_attrition_output_accuracy: 0.7873
Epoch 3/100
26/26 [==============================] - 0s 8ms/step - loss: 1.1622 - department_output_loss: 0.7506 - attrition_output_loss: 0.4115 - department_output_accuracy: 0.6663 - attrition_output_accuracy: 0.838

In [19]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_scaled, {"department_output": department_test_encoded, "attrition_output": attrition_test_encoded})
test_results

12/12 [==============================] - 0s 5ms/step - loss: 4.1440 - department_output_loss: 2.7713 - attrition_output_loss: 1.3727 - department_output_accuracy: 0.5543 - attrition_output_accuracy: 0.8179


[4.143996715545654,
 2.7712981700897217,
 1.3726980686187744,
 0.554347813129425,
 0.8179348111152649]

In [20]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {test_results[3]}")
print(f"Attrition Accuracy: {test_results[4]}")

Department Accuracy: 0.554347813129425
Attrition Accuracy: 0.8179348111152649


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. The attrition accuracy is the better metric to use based on the higher score of `0.8179`. Attrition is also a good metric for the problem being asked.

2. The activation function I used for Department was `softmax` because it's better suited for multi-class classification. As for Attrition, `sigmoid` is much better suited for binary tasks.

3. The model can be improved by optimizing parameters using GridSearch for example. Feature engineering is another viable way to increase accuracy, by providing more details on why employees left the company the model can gain more insight.